p2p/discovery/refresher.go

```diff
@@ -0,0 +1,218 @@
+package discovery
+
+import (
+	"context"
+	"errors"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/config"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"time"
+)
+
+// XXX TODO: move this impl to the upper protocol struct
+
+const (
+
+	// minTimeBetweenQueries is a minimum time between attempts to query a peer.
+	minTimeBetweenQueries = 500 * time.Millisecond
+	// lastQueriesCacheSize is the maximum size of the query cache map
+	lastQueriesCacheSize = 100
+)
+
+// ErrBootAbort is returned when when bootstrap is canceled by context cancel
+var ErrBootAbort = errors.New("bootstrap canceled by signal")
+
+// refresher is used to bootstrap and refresh peers in the addrbook
+type refresher struct {
+	logger log.Log
+	config config.SwarmConfig
+
+	bootNodes []discNode
+
+	book *addrBook
+
+	lastQueries map[p2pcrypto.PublicKey]time.Time
+
+	disc Protocol
+
+	quit chan struct{}
+}
+
+func newRefresher(book *addrBook, disc Protocol, config config.SwarmConfig, logger log.Log) *refresher {
+	bn := make([]discNode, 0, len(config.BootstrapNodes))
+	for _, n := range config.BootstrapNodes {
+		nd, err := node.NewNodeFromString(n)
+		if err != nil {
+			// TODO : handle errors
+			continue
+		}
+		bn = append(bn, discNodeFromNode(nd, nd.Address()))
+	}
+
+	//todo: trigger refresh every X with random nodes
+
+	return &refresher{
+		logger:      logger,
+		book:        book,
+		disc:        disc,
+		bootNodes:   bn,
+		lastQueries: make(map[p2pcrypto.PublicKey]time.Time),
+		quit:        make(chan struct{}), // todo: context ?
+	}
+}
+
+func (r *refresher) Bootstrap(ctx context.Context, num int) error {
+	var err error
+	tries := 0
+	servers := make([]discNode, 0, len(r.bootNodes))
+loop:
+	for {
+		size := r.book.NumAddresses()
+		if size == 0 {
+			r.book.AddAddresses(r.bootNodes, r.book.localAddress)
+			servers = r.bootNodes
+		}
+
+		if size-len(r.bootNodes) < num {
+			r.logger.Info("Bootstrap: starting with %v sized table", size)
+			res := r.refresh(servers)
```

Comment:
refresh does not imply anything about what this function actually does. it gets more peers so i'd call it something like `refreshPeerList` or something